Small updates to slice, str, io docs

bluss opened the pull request:
Main motivation was to update docs for the removal or "demotion" of certain extension traits. The update to the slice docs was larger, since the text was largely outdated.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nikomatsakis (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTING.md](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

bluss commented:
Point of discussion: Instead of saying `StrExt::split` it now refers to `str::split`. This is actually UFCS-callable which is really nice. The type vs module confusion might otherwise make it a bit confusing, but it's meant to be read as "the method split on type str."

steveklabnik commented on src/libcollections/slice.rs:
````diff
@@ -40,49 +39,44 @@
-//! * Further iterators exist that split, chunk or permute the slice.
+//! The mutable slice yields mutable references to the elements:
+//!
+//! ```rust
````
covention is to not include `rust` in API docs like this. I know it's strewn about with the rest of the file, but we shouldn't introduce new ones

steveklabnik commented on src/libcollections/slice.rs:
````diff
@@ -40,49 +39,44 @@
+//! ```
+//!
+//! The iterator yields references to the slice's elements, so if the element
+//! type of the slice is `i32`, the element type of the iterator is `&i32`.
````
Wouldn't these be `&mut i32`?

steveklabnik commented on src/libcollections/slice.rs:
```diff
@@ -40,49 +39,44 @@
+//! The iterator yields references to the slice's elements, so if the element
+//! type of the slice is `i32`, the element type of the iterator is `&i32`.
+//!
+//!
```
extra newline

steveklabnik commented on src/libcore/str/mod.rs:
```diff
@@ -147,8 +147,8 @@ impl FromStr for bool {
-    /// Note, in many cases, the StrExt::parse() which is based on
-    /// this FromStr::from_str() is more proper.
+    /// Note, in many cases, the `str::parse()` which is based on
+    /// this `FromStr::from_str()` is more proper.
```
I know you didn't write this sentence, but while you're here, maybe update it? It sounds akward to me:

> Note, in many cases, the str::parse() function is more proper.

steveklabnik commented:
Thank you! this is pretty great, r=me after we handle the nits

nikomatsakis commented:
r? @steveklabnik

nikomatsakis commented:
(I was just changing the assignment)

bluss commented:
Thank you for the review, fixed the nits! I changed the parse sentence to call it a method though.

steveklabnik commented:
Would you mind squashing? sorry

bluss commented:
Squashed into three commits again.

bors commented:
:umbrella: The latest upstream changes (presumably #23654) made this pull request unmergeable. Please resolve the merge conflicts.

bluss commented:
Rebased to new master.

Manishearth commented:
@bors: r=steveklabnik rollup

Looks like everything was addressed. Thanks!

bors commented:
:pushpin: Commit 547a48e has been approved by `steveklabnik`

<!-- @bors r=steveklabnik 547a48e1936ec7f369bf338fd7f5048f47265868 -->
